xwiki-platform-core/xwiki-platform-icon/xwiki-platform-icon-ui/src/main/resources/IconThemesCode/IconPicker.xml

```diff
@@ -286,29 +293,45 @@ require(['jquery', 'xwiki-icon-picker'], function($) {
   }
   
   /**
-   * Load the icon list (get the JSON from the server) and display it afterwards
+   * Load part of the icon list (get the JSON from the server) and display it afterwards
    */
   var loadIconList = function(iconTheme) {
-    let pageSize = 20;
-    let maxItems = 0;
-    $.getJSON(getResourceURL('data_iconamount', 'iconTheme=' + iconTheme) , function(dataIconAmount) {maxItems = 
-        dataIconAmount});
-    $.getJSON(getResourceURL('data_icons', 'iconTheme=' + iconTheme + '&number=' + pageSize.toString()), 
-    function(dataIcons) {
-      // Put the result in the icons map
-      icons[iconTheme] = dataIcons;
-      // Display the list
-      displayList(icons[iconTheme]);
-      // Hide the spinner
-      spinner.hide();
-    });
-    // Load more items if the user scrolled at the end of the current list and everything is not loaded yet.
-    iconListSection.off('scroll');
-    iconListSection.on('scroll', event => {
-      if ( iconListSection[0].scrollTop + iconListSection[0].clientHeight >= iconListSection[0].scrollHeight) {
-        console.log('I am on the bottom!');
+    var pageSize = 20;
+    var maxItems = -1;
+    // Recursive function helper. This calls the page to retrieve the icon information, by chunk of pageSize.
+    // After receiving the icon info, it displays them and calls itself again if the colortheme is not fully loaded.
+    var partialLoadIcon = function () {
+      let itemCount = icons[iconTheme].length;
+      if( maxItems === -1) {
+        // The first call will retrieve the total count of items we'll need to get.
+        $.getJSON(getResourceURL('data_icon_count', 'iconTheme=' + iconTheme) , function(dataIconCount) {
+          maxItems = dataIconCount
+        });
       }
-    });
+      $.getJSON(getResourceURL('data_icons', 'iconTheme=' + iconTheme
+        + '&amp;offset=' + itemCount.toString()
+        + '&amp;limit=' + pageSize.toString()),
+        function(dataIcons) {
+        // Put the result in the icons map
+        icons[iconTheme].push.apply(icons[iconTheme], dataIcons);
+        // Display the list of new icons.
+        // Since this recursive calls can take a while on poor network, we want to make sure we only display the newly
+        // retrieved icons if the iconTheme input still has their value.
+        if (iconTheme === currentIconTheme) {
+          displayList(dataIcons);
+        }
+        // Whatever the value of the current icon theme, we finish loading this one so that we have a correct state 
+        // if the user ever comes back to it.
+        if (itemCount + pageSize &lt; maxItems) {
```

Comment:
There's no guarantee that ``maxItems`` was updated by the first request when this code is called. You have two HTTP requests sent one after another, but there's no guarantee that the response will come in the order the requests were made. It will happen most of the time, but when it doesn't the icon theme will remain partially loaded. I think you should wait for the icon count to be retrieved before starting to fetch the icons (e.g. make ``partialLoadIcon`` async and use ``await``), or you could also always include the total count in the response that returns the icons.